cmd/pj-rehearse/main.go

```diff
@@ -1,23 +1,278 @@
 package main
 
 import (
+	"encoding/json"
 	"flag"
 	"fmt"
-	"log"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	// TODO: Solve this properly
+	"github.com/getlantern/deepcopy"
+	"github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+
+	pjapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	pjclientset "k8s.io/test-infra/prow/client/clientset/versioned"
+	pj "k8s.io/test-infra/prow/client/clientset/versioned/typed/prowjobs/v1"
+	prowconfig "k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/pjutil"
+
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/openshift/ci-operator-prowgen/pkg/diffs"
 )
 
+func getPrNumber(jobSpec *pjapi.ProwJobSpec) int {
+	return jobSpec.Refs.Pulls[0].Number
+}
+
+func getJobSpec() (*pjapi.ProwJobSpec, error) {
+	specEnv := []byte(os.Getenv("JOB_SPEC"))
+	if len(specEnv) == 0 {
+		return nil, fmt.Errorf("JOB_SPEC not set or set to an empty string")
+	}
+	spec := pjapi.ProwJobSpec{}
+	if err := json.Unmarshal(specEnv, &spec); err != nil {
+		return nil, err
+	}
+
+	if len(spec.Refs.Pulls) > 1 {
+		return nil, fmt.Errorf("Cannot rehearse in the context of a batch job")
+	}
+
+	return &spec, nil
+}
+
+func loadClusterConfig() (*rest.Config, error) {
+	clusterConfig, err := rest.InClusterConfig()
+	if err == nil {
+		return clusterConfig, nil
+	}
+
+	credentials, err := clientcmd.NewDefaultClientConfigLoadingRules().Load()
+	if err != nil {
+		return nil, fmt.Errorf("could not load credentials from config: %v", err)
+	}
+
+	clusterConfig, err = clientcmd.NewDefaultClientConfig(*credentials, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not load client configuration: %v", err)
+	}
+	return clusterConfig, nil
+}
+
+// Rehearsed Prow jobs may depend on ConfigMaps with content also modified by
+// the tested PR. All ci-operator-based jobs use the `ci-operator-configs`
+// ConfigMap that contains ci-operator configuration files. Rehearsed jobs
+// need to have the PR-version of these files available. The following code
+// takes care of creating a short-lived, rehearsal ConfigMap. The keys needed
+// to be present are extracted from the rehearsal jobs and the rehearsal jobs
+// are modified to use this ConfigMap instead of the "production" one.
+
+var ciOperatorConfigsCMName = "ci-operator-configs"
+
+type rehearsalCIOperatorConfigs struct {
+	cmclient  corev1.ConfigMapInterface
+	prNumber  int
+	configDir string
+
+	logger logrus.FieldLogger
+	dry    bool
+
+	configMapName string
+	neededConfigs map[string]string
+}
+
+func newRehearsalCIOperatorConfigs(cmclient corev1.ConfigMapInterface, prNumber int, configDir string, logger logrus.FieldLogger, dry bool) *rehearsalCIOperatorConfigs {
```

Comment:
If this is an unexported constructor, the callers could just set the fields themselves

Reply:
Right, thanks. Evolution artifact :)